ls1intum/Artemis
# Improve plagiarism detection for programming exercises

### Checklist

### Motivation and Context
This PR adds the possibility to inspect the plagiarism detection results for programming exercises in the Artemis client.

### Description
Currently, improved plagiarism detection is available for text and modeling exercises only. This PR adds the new plagiarism workflow to programming exercises, making it easier for instructors to inspect and compare similar programming submissions.

For now, we want to keep the possibility to download the report generated by JPlag. To generate and download the JPlag report instead of using the improved plagiarism inspection workflow, users can check the option 'Generate JPlag Report' before clicking on 'Check Plagiarism'.

### Steps for Testing

To properly test this PR, you need a programming exercise with at least two similar submissions.

1. Log in to Artemis
2. Navigate to Course Administration
3. Select a course
4. Go to Exercises
5. Navigate to the details page of a programming exercise by clicking 'View'
6. Click 'Check Plagiarism'

You should now see the plagiarism page for programming exercises. Make sure the following works:

* Clicking on 'Check Plagiarism' triggers the plagiarism detection and displays the result when it's done
* For programming exercises, the option 'Generate JPlag Report' should be visible and unchecked by default
* After the results are available, the first match is selected and the file contents of both submissions are displayed correctly
* Clicking on the split pane header (currently selected file) opens the file list for the given submission
* Selecting a file from the file list updates the split pane content accordingly
* When the option  'Generate JPlag Report' is checked, clicking on 'Check Plagiarism' should download a ZIP file with the JPlag report instead

### Test Coverage




### Screenshots
<img width="1680" alt="Screenshot 2020-12-16 at 13 16 05" src="https://user-images.githubusercontent.com/21973908/102347698-0e789480-3fa1-11eb-9815-9b9a9193ec66.png">
<img width="1680" alt="Screenshot 2020-12-16 at 13 16 23" src="https://user-images.githubusercontent.com/21973908/102347711-120c1b80-3fa1-11eb-98d2-58bf3d388c7d.png">
